### ISSUE TITLE
Type inference needs help with lambdas

### ISSUE THREAD

**Issue by MichalPokorny** [python/mypy#859]
This code:
```
from typing import TypeVar, Iterable, Iterator, Callable, Tuple
import itertools

_T = TypeVar('_T')

def f(iterable): # type: (Iterable[_T]) -> Iterator[List[_T]]
    grouped = itertools.groupby(enumerate(iterable), lambda pair: pair[0] // 2)
    return ([elem for _, elem in group] for _, group in grouped)
```
Gives: `test.py: note: In function "f": test.py:7: error: Value of type object is not indexable`.
Annotating the lambda with a concrete type seems to help:
```
def f(iterable): # type: (Iterable[_T]) -> Iterator[List[_T]]
    l = lambda pair: pair[0] // 2 # type: Callable[[Tuple[int, _T]], int]
    grouped = itertools.groupby(enumerate(iterable), l)
    return ([elem for _, elem in group] for _, group in grouped)
```

**Comment by JukkaL** [comment]
This may be a known issue with mypy getting confused when type inference result has type variables. I'll look at this later.

**Comment by gvanrossum** [comment]
I've got a simpler repro, it seems if the lambda is forward-referenced it doesn't have the type. E.g. this is an error:
```
def main() -> None:
    x = foo()
foo = lambda: 42
```
But it is silent if the definition of foo comes before main(), or if it is replaced with a def.

**Comment by JukkaL** [comment]
The latter issue is related to #1016 and multipass type checking would fix it.

**Comment by gvanrossum** [comment]
Oh wow. Much subtler.

**Comment by rwbarton** [comment]
The original program in this issue works now, but I wasn't able to easily determine why. It may even have been a typeshed change, I suppose.

**Comment by rwbarton** [comment]
Also, Guido's example works too now, presumably thanks to the fix for #1016.
